packages/cron/src/jobs/pins-backup.js

```diff
@@ -0,0 +1,301 @@
+    this.GET_PINNED_PINS_QUERY = `
+      SELECT DISTINCT ON(psa.id)
+        psa.id,
+        psa.backup_urls,
+        psa.source_cid,
+        psa.content_cid,
+        psa.name,
+        psa.auth_key_id,
+        pl.ipfs_peer_id
+      FROM psa_pin_request psa
+        JOIN pin p ON p.content_cid = psa.content_cid
+        JOIN pin_location pl ON pl.id = p.pin_location_id
+      WHERE p.status = 'Pinned'
+        AND psa.backup_urls = '{}'
+      LIMIT $1
+    `
+    this.s3 = getS3Client(env)
+    /**
+     * @type { number }
+     */
+    this.CONCURRENCY = env.CONCURRENCY !== undefined ? env.CONCURRENCY : 10
+  }
+
+  /**
+   * @param {import('pg').Pool} db
+   */
+  registerBackup (db, contentCid, pinRequestId) {
+    /**
+     * @param {AsyncIterable<import('../types/pins-backup').RemoteBackup>} source
+     */
+    return async (source) => {
+      for await (const bak of source) {
+        this.log(`⏳ Saving backup url to DB for ${contentCid}`)
+        await db.query(this.UPDATE_BACKUP_URL_QUERY, [
+          [bak.backupUrl.toString()],
+          pinRequestId,
+          contentCid.toString()
+        ])
+        this.log(`✅ Saved backup record for upload ${bak.contentCid}: ${bak.backupUrl.toString()}, rowId: ${pinRequestId}`)
+      }
+    }
+  }
+
+  /**
+   * @param {import('@aws-sdk/client-s3').S3Client} s3
+   * @param {string} bucketName
+   */
+  uploadCar (s3, bucketName) {
+    // rebind this for curried function
+    const _this = this
+
+    /**
+     * @param {AsyncIterable<import('../types/pins-backup').BackupContent} source
+     */
+    return async function * (source) {
+      for await (const bak of source) {
+        const backupUrl = await _this.s3Upload(s3, bucketName, bak)
+        /** @type {import('../types/pins-backup').RemoteBackup} */
+        const backup = { ...bak, backupUrl }
+        yield backup
+      }
+    }
+  }
+
+  /**
+   * @param {import('@aws-sdk/client-s3').S3Client} s3
+   * @param {string} bucketName
+   * @param {import('../types/pins-backup').BackupContent} bak
+   */
+  async s3Upload (s3, bucketName, bak) {
+    const key = `complete/${bak.sourceCid}.car`
+    const region = this.env.S3_BUCKET_REGION
+    const url = new URL(`https://${bucketName}.s3.${region}.amazonaws.com/${key}`)
+
+    try {
+      // Request the head object of the file we are about to backup
+      // If it throws a NotFound error then we know we need to upload it
+      const command = new HeadObjectCommand({ Bucket: bucketName, Key: key })
+      const res = await this.s3.send(command)
+      this.log('ℹ️ Car file already exists in S3. Skipping upload.', res)
+    } catch (err) {
+      if (err.name === 'NotFound') {
+        this.log(`⏳ Uploading to ${url}`)
+
+        const upload = new Upload({
+          client: s3,
+          params: {
+            Bucket: bucketName,
+            Key: key,
+            Body: bak.content,
+            Metadata: { structure: 'Complete' }
```

Comment:
Look into sending checksum for file, reject the upload if the bytes of car don't match the cid

Reply:
I wonder if this is required.
Looking at the headers sent by the client 
```
  'content-type': 'application/xml',
  'content-length': '11985',
  Expect: '100-continue',
  host: '127.0.0.1',
  'x-amz-user-agent': 'aws-sdk-js/3.53.1',
  'user-agent': 'aws-sdk-js/3.53.1 os/darwin/21.6.0 lang/js md/nodejs/16.14.0 md/crt-avail api/s3/3.53.1',
  'amz-sdk-invocation-id': '25110079-acaf-425f-8933-3527fd8366c7',
  'amz-sdk-request': 'attempt=1; max=3',
  'x-amz-date': '20221031T122520Z',
  'x-amz-content-sha256': 'ebd8a0f42b66a7756aaee73e6275d918143525f125137b584e6b079b364a6b5f',
  authorization: 'AWS4-HMAC-SHA256 Credential=minioadmin/20221031/us-east-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-user-agent, Signature=90453c633c07234480f3319eb3c1b058d25b39a077eb3653063165c5bc137722'
}
```

you can see it sends `x-amz-content-sha256` which is the sha256 of the payload and implies the payload is signed (see [docs](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html).

If every chunk is hashed and verified I don't think we need the overall one? Or am I missing something?

Reply:
I wonder if this is required.
Looking at the headers sent by the client 
```
  'content-type': 'application/xml',
  'content-length': '11985',
  Expect: '100-continue',
  host: '127.0.0.1',
  'x-amz-user-agent': 'aws-sdk-js/3.53.1',
  'user-agent': 'aws-sdk-js/3.53.1 os/darwin/21.6.0 lang/js md/nodejs/16.14.0 md/crt-avail api/s3/3.53.1',
  'amz-sdk-invocation-id': '25110079-acaf-425f-8933-3527fd8366c7',
  'amz-sdk-request': 'attempt=1; max=3',
  'x-amz-date': '20221031T122520Z',
  'x-amz-content-sha256': 'ebd8a0f42b66a7756aaee73e6275d918143525f125137b584e6b079b364a6b5f',
  authorization: 'AWS4-HMAC-SHA256 Credential=minioadmin/20221031/us-east-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-user-agent, Signature=90453c633c07234480f3319eb3c1b058d25b39a077eb3653063165c5bc137722'
}
```

you can see it sends `x-amz-content-sha256` which is the sha256 of the payload and implies the payload is signed (see [docs](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html).

If every chunk is hashed and verified I don't think we need the overall one? Or am I missing something?

Reply:
cc @olizilla